File: algorithms/sort/insertion-sort/insertion_sort.c
#include <stdlib.h>
#include "../commons/head/random.h"

static void insertion(int *arr, int length);

void insertion(int *arr, int length)
{
    // 临时保存待插入元素
    int temp;
    //指针 指向有序数组最后一个元素
    int *index;

    for (int i = 1; i < length; i++)
    {
        if (arr[i] > arr[i - 1])
            continue;

        index = &arr[i - 1];
        temp = arr[i];

        while (index >= &arr[0] && temp < *index)
        {
            *(index + 1) = *index;
            index--;
        }
        *(index+1) = temp;
        show(arr, length);
    }
}
// gcc -o run.out insertion_sort.c ../commons/source/random.c
// ./run.out 10
int main(int argc, const char *argv[])
{
    int length = atoi(argv[1]);
    int *mock = randomArray(length);
    insertion(mock, length);
    free(mock);
}

File: algorithms/sort/insertion-sort/insertion_sort.go
package main

import (
	"fmt"
	"os"
	"strconv"

	"bougainvilleas.org/commons"
)

func main() {
	length, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic(err)
	}
	mock := commons.RandomArray(length)
	insertion(mock, length)
}

func insertion(arr []int, length int) {
	var index int
	var temp int
	for i := 1; i < length; i++ {
		if arr[i] > arr[i-1] {
			continue
		}
		temp = arr[i]
		index = i - 1
		for index >= 0 && temp < arr[index] {
			arr[index+1] = arr[index]
			index--
		}
		arr[index+1] = temp
		fmt.Println(arr)
	}
}
